Conversions: Add conversions for grains, carats, drams, cc's, cubic inches, mmHg, decades, centuries, millennia, and thousandths of an inch

zblair opened the pull request:
###### What does your Pull Request do (check all that apply)?
- [x] Improvement
    - [x] Enhancement: **`Conversions: Add conversions for grains, carats, drams, cc's, cubic inches, mmHg, decades, centuries, millennia, and thousandths of an inch`**

###### Description of changes

Adds conversions for grains, carats, drams, cc's, cubic inches, mmHg, torr, decades, centuries, millennia, and thousandths of an inch

Grain: https://en.wikipedia.org/wiki/Grain_%28unit%29
Carat: https://en.wikipedia.org/wiki/Carat_%28mass%29
Dram: https://en.wikipedia.org/wiki/Dram_%28unit%29#Modern_unit_of_mass
Cubic centimetre: https://en.wikipedia.org/wiki/Cubic_centimetre
Cubic inches: https://en.wikipedia.org/wiki/Cubic_inch
mmHg: https://en.wikipedia.org/wiki/Millimeter_of_mercury
Torr: https://en.wikipedia.org/wiki/Torr

###### Which issues (if any) does this fix?

None

###### People to notify

@mintsoft

Instant Answer Page: https://duck.co/ia/view/conversions

[Maintainer](http://docs.duckduckhack.com/maintaining/guidelines.html): @mintsoft

daxtheduck commented:
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

GuiltyDolphin commented on share/goodie/conversions/ratios.yml:
```diff
@@ -678,6 +752,19 @@ type: pressure
+  - torrs
+factor: 0.00750062
+type: pressure
+unit: torr
```
We should add some tests for the new units.

mintsoft commented:
@zblair Thanks for this; unfortunately the tests are failing. I'm just seeing if I can figured out why

mintsoft commented on share/goodie/conversions/ratios.yml:
```diff
@@ -142,6 +168,17 @@ type: length
+  - thousandths of an inch
+factor: 39370.1
+type: length
+unit: thousandth of an inch
```
You'll need to add this to the plural exceptions: https://github.com/zblair/zeroclickinfo-goodies/blob/master/lib/DDG/Goodie/Conversions.pm#L39

mintsoft commented on share/goodie/conversions/ratios.yml:
```diff
@@ -611,6 +664,27 @@ type: duration
+  - millennia
+factor: 0.00000273972602739726
+type: duration
+unit: millennium
```
Same here: https://github.com/zblair/zeroclickinfo-goodies/blob/master/lib/DDG/Goodie/Conversions.pm#L39

mintsoft commented on share/goodie/conversions/ratios.yml:
```diff
@@ -611,6 +664,27 @@ type: duration
+  - centuries
+factor: 0.0000273972602739726
+type: duration
+unit: century
```
and here: https://github.com/zblair/zeroclickinfo-goodies/blob/master/lib/DDG/Goodie/Conversions.pm#L39

zblair commented:
Thank you for the review comments, @mintsoft and @GuiltyDolphin! I think that I have addressed those comments with my previous two commits.

I believe that I also noticed a bug. Search queries like "5 pas.i to atm" or "134 pasti to pascals" would cause the script to die.

This was because "pas.i" and "pasti" would incorrectly match the "p.s.i" alias in ratios.yml. The problem was that "p.s.i" was interpreted as a regex where the "." characters would be treated as wildcards.

My solution was to process each alias in ratios.yml using the `quotemeta` to escape regex-reserved characters.

GuiltyDolphin commented on t/Conversions.t:
```diff
@@ -2657,6 +2777,7 @@ ddg_goodie_test(
     '99999999999000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000 miles in mm' => undef,
-    '1E300 miles in metres'           => undef
+    '1E300 miles in metres'           => undef,
+    '5 pas.i to atm'                  => undef
```
May as well stick a comma on the end here so we don't get weird diffs later on.

zblair commented:
Thanks, I have added the trailing comment now.

mintsoft commented:
Excellent; thanks for the improvement and fixes! @zblair
